File: 0101-0200/0164.Maximum_Gap.java
//40ms





class Solution {
    public int maximumGap(int[] nums) {
        if (nums == null || nums.length < 2){
            return 0;
        }
        
        Arrays.sort(nums);
        int maxDiff = 0;
        for (int i = 1; i < nums.length; i++) {
            int diff = nums[i] - nums[i - 1];
            if (diff > maxDiff) {
                maxDiff = diff;
            }
        }
        return maxDiff;
    }
}







//7ms







class Solution {
    public int maximumGap(int[] nums) {
        int n=nums.length;
        if(n<2){
            return 0;
        }
        if(n==2){
            return Math.abs(nums[1]-nums[0]);
        }
        int res= bucketSort(nums);
        return res;
    }

    public int bucketSort(int[] arr){
        int n=arr.length;
        int max = 0, min = Integer.MAX_VALUE;

        for(int i: arr){
            if(i>max) max =i;
            if(i<min) min=i;
        }
        if(max == min)
            return 0;
        
        int bucketSize = (max - min)/n + 1;  
        int bucketCount = (max-min)/bucketSize + 1;
        int[] minBucket = new int[bucketCount];
        int[] maxBucket = new int[bucketCount];

        Arrays.fill(minBucket,Integer.MAX_VALUE);
        for(int x:arr){
            int index = (x - min) / bucketSize;
            minBucket[index] = Math.min(minBucket[index], x);
            maxBucket[index] = Math.max(maxBucket[index], x);
        }

        int prevMax = maxBucket[0], result = 0;
        for(int i = 1; i < bucketCount; ++i) {
            if(minBucket[i] == Integer.MAX_VALUE)
                continue;
            result = Math.max(result, minBucket[i] - prevMax);
            prevMax = maxBucket[i];
        }
        return result;
    }
}







//7ms







public class Solution {
    public int maximumGap(int[] nums) {
        if (nums == null || nums.length < 2) {
            return 0;
        }
        int n = nums.length;
        int min = Integer.MAX_VALUE;
        int max = Integer.MIN_VALUE;
        for (int num : nums) {
            min = Math.min(min, num);
            max = Math.max(max, num);
        }
        if (min == max) {
            return 0;
        }
        int bucketSize = (int) Math.ceil((double) (max - min) / (n - 1));
        int bucketCount = (max - min) / bucketSize + 1;
        int[] bucketMin = new int[bucketCount];
        int[] bucketMax = new int[bucketCount];
        for (int i = 0; i < bucketCount; i++) {
            bucketMin[i] = Integer.MAX_VALUE;
            bucketMax[i] = Integer.MIN_VALUE;
        }
        for (int num : nums) {
            int bucketIndex = (num - min) / bucketSize;
            bucketMin[bucketIndex] = Math.min(bucketMin[bucketIndex], num);
            bucketMax[bucketIndex] = Math.max(bucketMax[bucketIndex], num);
        }
        int maxGap = 0;
        int prevMax = min;
        for (int i = 0; i < bucketCount; i++) {
            if (bucketMin[i] == Integer.MAX_VALUE) {
                continue;
            }
            maxGap = Math.max(maxGap, bucketMin[i] - prevMax);
            prevMax = bucketMax[i];
        }
        return maxGap;
    }
}

File: 0101-0200/0164.Maximum_Gap.js
//77ms






/**
 * @param {number[]} nums
 * @return {number}
 */
var maximumGap = function (nums) {
  if (nums.length <= 1) return 0

  nums = radixSort(nums)

  let maxDiff = Math.abs(nums[0] - nums[1])

  for (let i = 1; i < nums.length - 1; i++) {

    let currDiff = Math.abs(nums[i] - nums[i + 1])
    maxDiff = Math.max(maxDiff, currDiff)
  }

  return maxDiff
};

const radixSort = (nums) => {
  const mostDigit = mostDigits(nums)

  for (let k = 0; k < mostDigit; k++) {
    let buckets = Array.from({ length: 10 }, () => [])

    for (let i = 0; i < nums.length; i++) {
      let digit = getDigit(nums[i], k)

      buckets[digit].push(nums[i])
    }
    nums = [].concat(...buckets)
  }

  return nums
}

const getDigit = (num, place) => {
  return Math.floor(Math.abs(num) / Math.pow(10, place)) % 10
}

const digitCount = (num) => {
  if (num === 0) return 1

  return Math.floor(Math.log10(Math.abs(num))) + 1
}

const mostDigits = (nums) => {
  let mostDigitCount = 0

  for (let num of nums) {
    mostDigitCount = Math.max(mostDigitCount, digitCount(num))
  }

  return mostDigitCount
}






//33ms






/**
 * @param {number[]} nums
 * @return {number}
 */
var maximumGap = function(nums) {
    const n = nums.length
    if(n < 2){
        return 0
    }
    let maxGap = 0

    let min = Math.min(...nums)
    let max = Math.max(...nums)

    const bucketSize = Math.ceil((max-min)/(n-1))
    const minOfBucket = new Array(n-1).fill(Infinity)
    const maxOfBucket = new Array(n-1).fill(-1)

// set the min and max of bucket means put the value to the min and max bucket
    for(let i=0;i<n;i++){
        if(nums[i] === min || nums[i] === max){
            continue
        }
        const buckIndxForCurrEle = Math.floor((nums[i]-min)/bucketSize)
        minOfBucket[buckIndxForCurrEle] = Math.min(minOfBucket[buckIndxForCurrEle], nums[i])
        maxOfBucket[buckIndxForCurrEle] = Math.max(maxOfBucket[buckIndxForCurrEle], nums[i])
    }

// Now we get the diff with max of prev bucket and min of next bucket.
    for(let i=0;i<n-1;i++){
        if(maxOfBucket[i] === -1){
            continue
        }
        maxGap = Math.max(maxGap, minOfBucket[i] - min)
        min = maxOfBucket[i]
    }
    return Math.max(maxGap, max-min)
};
